### ISSUE TITLE
Questions about the files used for indicator calculation

### ISSUE THREAD

**Issue by xiachashuanghua** [jianlin-cheng/CryoSegNet#12]
Hello!
I encountered some issues regarding indicator calculation while using your project. Specifically, I am not sure how indicator calculation is carried out in the project, especially when it comes to the use of Python files
Is evaluate.py used for indicator calculation? If this file is used, why is the SAM model not used

**Comment by rajangyawali** [comment]
Yes, the evaluation.py is used to calculate indicators.  You can also evaluate for other models using precision_recall.py. For this you need to generate masks for each of the models.

Thank you.

**Comment by xiachashuanghua** [comment]
> Yes, the evaluation.py is used to calculate indicators. You can also evaluate for other models using precision_recall.py. For this you need to generate masks for each of the models.
> 
> Thank you.
I tried to use evaluation.py in combination with predict.py to calculate the metrics of UNet+SAM. However, the obtained metrics are much lower than those described in the paper. Could you help me check what's wrong with the code? Or is there any other way for me to obtain the metrics in the paper without using the images in General?

```
# Code for calculating different evaluation metrics like Precision, Recall, F1-Score and Dice Score.
import sys
import os
current_dir = os.path.dirname(os.path.abspath(__file__))
parent_dir = os.path.dirname(current_dir)
sys.path.insert(0, parent_dir)
import copy
import config
import matplotlib.pyplot as plt

import numpy as np
import torch
import cv2
from dataset.dataset import transform
import glob
from models.model_5_layers import UNET
import config
import mrcfile
from segment_anything import SamAutomaticMaskGenerator, sam_model_registry
import statistics as st
print("[INFO] Loading up model...")
model = UNET().to(device=config.device)

sam_model = sam_model_registry[config.model_type](checkpoint=config.sam_checkpoint)
sam_model.to(device=config.device)

mask_generator = SamAutomaticMaskGenerator(sam_model)

def get_annotations(anns):
    if len(anns) == 0:
        return
    sorted_anns = sorted(anns, key=(lambda x: x['area']), reverse=True)
    ax = plt.gca()
    ax.set_autoscale_on(False)

    img = np.ones((sorted_anns[0]['segmentation'].shape[0], sorted_anns[0]['segmentation'].shape[1], 4))
    img[:,:,3] = 0
    for ann in sorted_anns:
        m = ann['segmentation']
        color_mask = np.concatenate([np.random.random(3), [0.35]])
        img[m] = color_mask
    
    return img

state_dict = torch.load('/root/autodl-tmp/pretrained_models/cryosegnet.pth') 
model.load_state_dict(state_dict, strict=True)


def evaluation_metrics(model, image_path, threshold):
    model.eval()
    with torch.no_grad():
        mask_path = image_path.replace("images", "masks")
        mask_path = mask_path.replace(".jpg", "_mask.jpg")   
        # image = mrcfile.read(image_path)
        # image = image.T
        # image = np.rot90(image)
        image = cv2.imread(image_path, 0)
        mask = cv2.imread(mask_path, 0)
        height, width = image.shape
        orig_image = copy.deepcopy(image)
        orig_mask = copy.deepcopy(mask)
        image = cv2.resize(image, (config.input_image_width, config.input_image_height))        
        mask = cv2.resize(mask, (config.input_image_width, config.input_image_height))
        # segment_mask = copy.deepcopy(orig_image)
        segment_mask = np.zeros((height, width), dtype=np.uint8)
        mask = mask/255
        orig_mask=orig_mask/255
        image = torch.from_numpy(image).unsqueeze(0).float()
        image = image / 255.0
        image = image.to(config.device).unsqueeze(0)
        
        predicted_mask = model(image)         
        predicted_mask = torch.sigmoid(predicted_mask)
        predicted_mask = predicted_mask.cpu().numpy().reshape(config.input_image_width, config.input_image_height)

        sam_output = np.repeat(transform(predicted_mask)[:,:,None], 3, axis=-1)
        predicted_mask = cv2.resize(predicted_mask, (width, height))
        
        masks = mask_generator.generate(sam_output)
        sam_mask = get_annotations(masks)
        sam_mask = cv2.resize(sam_mask, (width, height) )
        bboxes = []
        for i in range(0, len(masks)):
            if masks[i]["predicted_iou"] > 0.94:
                box = masks[i]["bbox"]
                bboxes.append(box)
        
        if len(bboxes) > 1:
        
            x_ = st.mode([box[2] for box in bboxes])
            y_ = st.mode([box[3] for box in bboxes])
            d_ = np.sqrt((x_ * width / config.input_image_width)**2 + (y_ * height / config.input_image_height)**2)
            r_ = int(d_//2)
            th = r_ * 0.20
            segment_mask = cv2.cvtColor(segment_mask, cv2.COLOR_GRAY2BGR)
            for b in bboxes:
                if b[2] < x_ + th and b[2] > x_ - th/3 and b[3] < y_ + th and b[3] > y_ - th/3: 
                    x_new, y_new = int((b[0] + b[2]/2) / config.input_image_width * width) , int((b[1] + b[3]/2) / config.input_image_height * height)
                    coords = cv2.circle(segment_mask, (x_new, y_new),  r_, (255, 255, 255), -1)
            try:
                smooth = 0.001  
                pred_mask = copy.deepcopy(coords)
                pred_mask = cv2.cvtColor(pred_mask, cv2.COLOR_BGR2GRAY)
                pred_mask=pred_mask/255
                # pred_mask[pred_mask >= threshold] = 1.0
                # pred_mask[pred_mask < threshold] = 0         

                true_positive = np.sum(np.logical_and(orig_mask, pred_mask))
                false_positive = np.sum(np.logical_and(np.logical_not(orig_mask), pred_mask))
                false_negative = np.sum(np.logical_and(orig_mask, np.logical_not(pred_mask)))

                # Calculate precision, recall, f1, dice score
                precision = (true_positive  + smooth) / (true_positive + false_positive + smooth)
                recall = (true_positive  + smooth) / (true_positive + false_negative + smooth)

                inputs = orig_mask.reshape(-1)
                targets = pred_mask.reshape(-1)

                intersection = (inputs * targets).sum()
                dice_score = (2. * intersection + smooth) / (inputs.sum() + targets.sum() + smooth) 

                return precision, recall, dice_score
            except:
                pass
        else:
            pass
        # predicted_mask = predicted_mask.cpu().numpy()
        


def evaluation(model, images_path, empiar_id, threshold=0.1):
    total_precisions = []
    total_recalls = []
    total_dice_scores = []

    for i in range(0, len(images_path), 1):
        precision, recall, dice_score = evaluation_metrics(model, images_path[i], threshold)
        total_precisions.append(precision)
        total_recalls.append(recall)
        total_dice_scores.append(dice_score)
        
    precision = np.round(np.max(total_precisions, axis = 0), 3)
    recall = np.round(np.max(total_recalls, axis = 0), 3)  
    f1_score = np.round((2 * precision * recall) / (precision + recall), 3)
    dice_score = np.round(np.max(total_dice_scores, axis = 0), 3) 
    
    print("\tPrecision", precision)
    print("\tRecall", recall)
    print("\tF1-Score", f1_score)
    print("\tDice Score", dice_score)
    

empiar_ids = [10028, 10081, 10345, 11056, 10532, 10093, 10017]
print("[INFO] Loading up test images path ...")
for empiar_id in empiar_ids:
    images_path = list(glob.glob(f"{config.test_dataset_path}/{empiar_id}/images/*.jpg"))
    print("\n")
    print(f"Evaluation Results for EMPIAR ID {empiar_id}")
    evaluation(model, images_path, empiar_id)
    print(f"--------------------------------------")
```

![Image](https://github.com/user-attachments/assets/070c1c39-2c56-4ef4-921e-8d8b988447d1)
